File: module-3-tools/solutions/python/tool_registry.py
"""Module 3: Tool Registry."""
from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass
class Tool:
    name: str
    description: str
    input_schema: dict[str, Any]
    execute: Callable[[dict[str, Any]], str]


class ToolRegistry:
    def __init__(self):
        self._tools: dict[str, Tool] = {}

    def register(self, tool: Tool) -> None:
        self._tools[tool.name] = tool

    def get_definitions(self) -> list[dict]:
        return [
            {"name": t.name, "description": t.description, "input_schema": t.input_schema}
            for t in self._tools.values()
        ]

    def execute(self, name: str, inp: dict[str, Any]) -> str:
        tool = self._tools.get(name)
        if not tool:
            return f'Error: unknown tool "{name}"'
        try:
            return tool.execute(inp)
        except Exception as e:
            return f"Error: {e}"

    def has(self, name: str) -> bool:
        return name in self._tools

    def list(self) -> list[str]:
        return list(self._tools.keys())

File: module-3-tools/solutions/tool-registry.ts
/**
 * Module 3: Tool Registry
 * Register tools, get definitions for API, execute with error wrapping.
 */
import type Anthropic from "@anthropic-ai/sdk";
import type { Tool } from "../../module-2-agent-loop/solutions/types.js";

export class ToolRegistry {
  private tools = new Map<string, Tool>();

  register(tool: Tool): void {
    this.tools.set(tool.name, tool);
  }

  getDefinitions(): Anthropic.Tool[] {
    return Array.from(this.tools.values()).map(({ name, description, input_schema }) => ({
      name,
      description,
      input_schema: input_schema as Anthropic.Tool["input_schema"],
    }));
  }

  async execute(name: string, input: Record<string, unknown>): Promise<string> {
    const tool = this.tools.get(name);
    if (!tool) return `Error: unknown tool "${name}"`;
    try {
      return await tool.execute(input);
    } catch (e) {
      return `Error: ${e instanceof Error ? e.message : String(e)}`;
    }
  }

  has(name: string): boolean {
    return this.tools.has(name);
  }

  list(): string[] {
    return Array.from(this.tools.keys());
  }
}
